From 3ab8313bb0a5f4d795f8af0130a48030a2c7b02d Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?=C3=96zg=C3=BCr=20Kele=C5=9F?= <ozgur.keles@ries.com.tr>
Date: Thu, 3 Dec 2015 11:09:35 +0200
Subject: [PATCH] 4-power-fail bug fix root_device_flush fonksiyonu eklendi.

---
 drivers/misc/power-fail.c | 25 +++++++++++++++++++++++++
 1 file changed, 25 insertions(+)

diff --git a/drivers/misc/power-fail.c b/drivers/misc/power-fail.c
index abf1a24..e3181c2 100644
--- a/drivers/misc/power-fail.c
+++ b/drivers/misc/power-fail.c
@@ -16,6 +16,7 @@
 #include <asm/atomic.h>
 #include <linux/slab.h>
 #include <linux/string.h>
+#include <linux/syscalls.h>
 
 #ifdef CONFIG_PLAT_SUNXI
 #include <plat/sys_config.h>
@@ -74,6 +75,7 @@ enum PowerFailBits
     , PowerFailIRQBit
     , PowerFailStartIRQ
     , PowerFailSuperCapEnable
+    , RootDeviceFlushBit
 
 
     , PowerFailBitCount
@@ -114,6 +116,27 @@ static void open_lcd(void)
     }
 }
 
+static void root_device_flush(void)
+{
+    int error = 0;
+    int root_fd = -1;
+    if (!test_and_set_bit(RootDeviceFlushBit, &powerfail_bits)) {
+        root_fd = sys_open("/dev/root", O_RDWR, 0);
+        if (root_fd < 0) {
+            error = root_fd;
+            printk(KERN_ERR DEVICE_NAME ": Flush Root Device Open Error: %d\n", error);
+        } else {
+            error = sys_ioctl(root_fd, BLKFLSBUF, 0);
+            sys_close(root_fd);
+            if(!error)
+                printk(KERN_INFO DEVICE_NAME ": Flush Root Device\n");
+            else
+                printk(KERN_ERR DEVICE_NAME ": Flush Root Device ioctl Error %d\n", error);
+        }
+        clear_bit(RootDeviceFlushBit, &powerfail_bits)
+    }
+}
+
 static void start_timer_interrupt(void)
 {
     if (!test_and_set_bit(PowerFailStartIRQ, &powerfail_bits)) {
@@ -341,6 +364,8 @@ static ssize_t powerfail_write(struct file *filp, const char __user *user_buffer
                 enable_super_cap();
             } else if (strcmp(buffer, "disable_super_cap") == 0) { //disable super cap
                 disable_super_cap();
+            } else if (strcmp(buffer, "root_device_flush") == 0) { //root device flush
+                root_device_flush();
             }
             retval = count;
         }
-- 
2.6.0

