From 9d564f0fcac29e4e7d87848c7a2a60086d8d4aa8 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?=C3=96zg=C3=BCr=20Kele=C5=9F?= <ozgur.keles@ries.com.tr>
Date: Tue, 9 Dec 2014 16:56:27 +0200
Subject: [PATCH] bug fix sun4i-keypad

---
 drivers/input/keyboard/sun4i-keypad.c | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/drivers/input/keyboard/sun4i-keypad.c b/drivers/input/keyboard/sun4i-keypad.c
index 21fc478..266eff8 100644
--- a/drivers/input/keyboard/sun4i-keypad.c
+++ b/drivers/input/keyboard/sun4i-keypad.c
@@ -406,13 +406,13 @@ static int __devinit sw_keypad_probe(struct platform_device *pdev)
 	}
     disable_irq(keypad->irq);
 
+	platform_set_drvdata(pdev, keypad);
+	keypad->pdev = pdev;
+
 	error = input_register_device(keypad->input_dev);
 	if (error)
 		goto err_free_irq;
 
-	platform_set_drvdata(pdev, keypad);
-	keypad->pdev = pdev;
-
 	swkp_msg("sw keypad probe done, base %p, irq %d\n", keypad->base, keypad->irq);
 	return 0;
 
@@ -534,7 +534,7 @@ static int __init sw_keypad_init(void)
 
     swkp_msg("sw keypad init\n");
     kp_used  = 0;
-    ret = script_parser_fetch("keypad_para", "ke_used", &kp_used, sizeof(int));
+    ret = script_parser_fetch("keypad_para", "kp_used", &kp_used, sizeof(int));
     if (ret)
     {
         printk("sw keypad fetch keypad uning configuration failed\n");
-- 
2.1.3

